[yarnmengnus] iP

yarnmengnus opened the pull request:
### **DUKE** 
_a module deliverable_ 
`a Personal Assistant Chatbot that helps a person to keep track of various things.` - Project Duke ([source](https://nus-cs2103-ay2324s1.github.io/website/se-book-adapted/projectDuke/index.html))
**Features**
- Text-based
- Easy to use

**How to install**
1.  Download it from [here](https://github.com/yarnmengnus/ip/releases/download/A-Jar/duke.jar)
2.  Double click it
3. Add you tasks! 👍 

**Capabilities** 
- [X] Add Events, Deadlines, Todos
- [X] Find your tasks by name
- [ ] Sorting tasks by time 
- [ ] Reminders

And its **OPEN SOURCED**. Find the repo [here](https://github.com/yarnmengnus/ip).

Here's a sample of the `main` method.
```
public static void main(String[] args) {
        Ui ui = new Ui();
        Storage storage = new Storage(DukeEnvironmentConstants.FILE_PATH);
        Parser parser = new Parser();
        ChatBotList list = new ChatBotList(storage.loadList());
        System.out.println(new File(DukeEnvironmentConstants.FILE_PATH));
        while (true) {
            String input = ui.getNextLine().trim();
            try {
                Command c = parser.parse(input);
                c.execute(list, ui, storage);
                if (c instanceof ExitCommand) {
                    break;
                }
            } catch (DukeException e) {
                ui.showException(e);
            }
        }
        ui.close();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

yarnmengnus committed:
added console.bat

yarnmengnus committed:
Level-0

yarnmengnus committed:
Level-1

yarnmengnus committed:
Level-2

yarnmengnus committed:
Level-3

yarnmengnus committed:
Level-3 (edited output msg and code format)

yarnmengnus committed:
Level-4

yarnmengnus committed:
A-TextUiTesting

yarnmengnus committed:
Level-5

yarnmengnus committed:
Level-6

yarnmengnus committed:
A-Enums

yarnmengnus committed:
Add functionality to save to disk

yarnmengnus committed:
Edit some messaging

yarnmengnus committed:
Merge branch 'branch-level-7'

yarnmengnus committed:
Level 7

yarnmengnus committed:
Level 8

yarnmengnus committed:
Merge branch 'branch-Level-8'

yarnmengnus committed:
Remove redundant Exception

yarnmengnus committed:
Remove empty Item.java

yarnmengnus committed:
Tidy up the file system.

yarnmengnus committed:
Add Parser and Commands

yarnmengnus committed:
A-MoreOOP

yarnmengnus committed:
Merge branch 'A-MoreOOP'

yarnmengnus committed:
Package into duke

yarnmengnus committed:
Merge branch 'branch-A-Packages'

yarnmengnus committed:
Merge remote-tracking branch 'origin/add-gradle-support'

yarnmengnus committed:
Add Gradle Support

yarnmengnus committed:
Added a test JUnit Test.`

yarnmengnus committed:
Add tests for Parser mark and unmark

yarnmengnus committed:
A-JUnit

yarnmengnus committed:
Merge branch 'branch-A-JUnit'

yarnmengnus committed:
Add JavaDocs

yarnmengnus committed:
Fix bug with Ui

yarnmengnus committed:
Align the code to style guide.

yarnmengnus committed:
Level-9

yarnmengnus committed:
Merge branch 'branch-Level-9'

yarnmengnus committed:
Merge branch 'branch-A-JavaDoc'

yarnmengnus committed:
Merge branch 'branch-A-CodingStandard'

Bombbird2001 commented on src/main/java/duke/IllegalChatBotListArgumentException.java:
```diff
@@ -0,0 +1,12 @@
+    public IllegalChatBotListArgumentException(String message) {
+        super(message);
+    }
+    public IllegalChatBotListArgumentException() {
```
Should there be an empty line before this line? Also noticed in a few other files

Bombbird2001 commented on src/main/java/duke/ChatBotList.java:
```diff
@@ -0,0 +1,122 @@
+     * @return The toString of the task
+     */
+    public String addToList(String[] queries, DukeEnvironmentConstants.TaskType type) {
+        switch (type) {
+        case EVENT:
+            this.list.add(new Event(queries[0],
+                    LocalDateTime.parse(queries[1], DukeEnvironmentConstants.FORMATTER1),
+                    LocalDateTime.parse(queries[2], DukeEnvironmentConstants.FORMATTER1)));
+            break;
+        case DEADLINE:
+            this.list.add(new Deadline(queries[0],
+                    LocalDateTime.parse(queries[1], DukeEnvironmentConstants.FORMATTER1)));
+            break;
+        case TODO:
+            this.list.add(new Todo(queries[0]));
+            break;
+        default:
+            break;
+        }
```
I like that these follow the switch-case style guidelines

Bombbird2001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,119 @@
+                        throw new ParserException(INVALID_DATE_FORMAT);
+                    }
+                    return new AddToListCommand(DukeEnvironmentConstants.TaskType.EVENT,
+                            new String[]{secondSplit[0], thirdSplit[0], thirdSplit[1]});
+                }
+            } else if (splitInput[0].equals("deadline")) {
+                if (splitInput.length < 2) {
+                    throw new ParserException(EMPTY_DEADLINE_DESC);
+                } else {
+                    String[] secondSplit = splitInput[1].split(" +/by +", 2);
+                    if (secondSplit.length != 2) {
+                        throw new ParserException("Incorrect syntax!\n"
+                                + "Please use this syntax:\n"
+                                + "deadline <desc> /by <deadline>");
+                    }
+                    try {
+                        LocalDateTime.parse(secondSplit[1], DukeEnvironmentConstants.FORMATTER1);
+                    } catch (DateTimeParseException e) {
+                        throw new ParserException(INVALID_DATE_FORMAT);
+                    }
+                    return new AddToListCommand(DukeEnvironmentConstants.TaskType.DEADLINE,
+                            new String[]{secondSplit[0], secondSplit[1]});
+                }
+            } else if (splitInput[0].equals("todo")) {
+                if (splitInput.length < 2) {
+                    throw new ParserException(EMPTY_TODO_DESC);
+                } else {
+                    return new AddToListCommand(DukeEnvironmentConstants.TaskType.TODO, new String[]{splitInput[1]});
+                }
+            } else if (splitInput[0].equals("find")) {
+                if (splitInput.length < 2) {
+                    throw new ParserException(EMPTY_FIND_QUERY);
+                } else {
+                    return new FindCommand(splitInput[1]);
+                }
+            } else {
+                throw new ParserException(NO_VALID_COMMAND);
+            }
+        }
+    }
```
Perhaps can consider splitting this method into separate methods for parsing each type of command? It could help reduce the length, nesting and improve readability

Bombbird2001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,119 @@
+ * Encapsulates the parsing mechanism of the chat bot.
+ */
+public class Parser {
+    private static final String EMPTY_MARK_QUERY = "☹ OOPS!!! "
+            + "You did not specify which task you want to %s.\n"
+            + "Please use this syntax:\n %s <index>\n";
+    private static final String EMPTY_EVENT_DESC = "☹ OOPS!!! "
+            + "The description of a event cannot be empty.\nPlease use this syntax:\n"
+            + "event <desc> /from <start> /to <end>\n";
+    private static final String EMPTY_DEADLINE_DESC = "☹ OOPS!!! "
+            + "The description of a deadline cannot be empty.\n"
+            + "Please use this syntax:\n"
+            + "deadline <desc> /by <deadline>\n";
+    private static final String EMPTY_TODO_DESC = "☹ OOPS!!! "
+            + "The description of a todo cannot be empty.\n";
+    private static final String NO_VALID_COMMAND = "☹ OOPS!!! "
+            + "I'm sorry, but I don't know what that means :-(\n";
+    private static final String INVALID_DATE_FORMAT = "☹ OOPS!!! "
+            + "Invalid date/time format.\n"
+            + "Please format your start/end times as \"yyyy-MM-dd HHmm\" in 24 hr format";
+    private static final String INVALID_DATE_ORDER = "☹ OOPS!!! "
+            + "Your end date/time is before your start date/time!";
+    private static final String EMPTY_FIND_QUERY = "☹ OOPS!!! "
+            + "What are you finding?";
```
I like the use of these variables to help reduce magic constants

Bombbird2001 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,119 @@
+                } else {
+                    String[] secondSplit = splitInput[1].split(" +/from +");
+                    if (secondSplit.length != 2) {
+                        throw new ParserException("Incorrect syntax!\n"
+                                + "Please use this syntax:\n"
+                                + "event <desc> /from <start> /to <end>");
```
Great that the line breaks follow the style guidelines

Bombbird2001 reviewed (commented):
LGTM, just a few nits here and there

yarnmengnus committed:
Implement small gui

yarnmengnus committed:
Add Enter button functionality

yarnmengnus committed:
Merge branch 'branch-Level-10'

nabonitasen commented on src/main/java/duke/ParserException.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Encapsulates an error thrown by the Parser.
+ */
+public class ParserException extends DukeException {
```
I like how you have separated the different kinds of DukeExceptions.

nabonitasen commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+     * Handles commands and messages to ChatBot
+     * @param args CLI args. No implementation.
+     */
+    // public static void main(String[] args) {
+    //     Ui ui = new Ui();
+    //     Storage storage = new Storage(DukeEnvironmentConstants.FILE_PATH);
```
Could removing commented out code perhaps make it neater?

nabonitasen commented on src/main/java/duke/Launcher.java:
```diff
@@ -0,0 +1,10 @@
+
+import duke.uiux.Ui;
+import javafx.application.Application;
+
+public class Launcher {
+    public static void main(String[] args) {
```
Does this require JavaDoc?

nabonitasen commented on src/main/java/duke/ExitCommand.java:
```diff
@@ -0,0 +1,16 @@
+package duke;
+
+/**
+ * Encapsulates the command for exiting the chat bot.
```
I like your comments - very straightforward and understandable.

nabonitasen reviewed (commented):
Looks good overall! Just minimal styling/documentation fixes.

yarnmengnus committed:
Implement a test for assert

yarnmengnus committed:
Implement more assertions

yarnmengnus committed:
Improve the code quality

yarnmengnus committed:
Merge pull request #3 from yarnmengnus/branch-A-Assertions

Implement more assertions

yarnmengnus committed:
Merge branch 'master' into branch-A-CodeQuality

yarnmengnus committed:
Merge pull request #2 from yarnmengnus/branch-A-CodeQuality

Improve the code quality

yarnmengnus committed:
Implements part of the snooze command

Before, to change a deadline, one has to delete it and create it again.
This commit adds parts of the implementation. Still a work in progress.

yarnmengnus committed:
Implement the snooze command

Previously this was still work in progress. This commit will implement
it fully. Users can now snooze deadlines for 1 day. If the requested
item is not a deadline, an error message will be shown.

yarnmengnus committed:
Merge branch 'branch-BCD-Extension'

yarnmengnus committed:
Fix bugs in snooze

yarnmengnus committed:
Improve the ReadMe

yarnmengnus committed:
Fix ReadMe image links

yarnmengnus committed:
Another attempt at fixing image links for ReadMe

yarnmengnus committed:
Attempt to fix images in ReadMe again

yarnmengnus committed:
Empty commit

yarnmengnus committed:
Attempt to image paths

yarnmengnus committed:
Change image name

yarnmengnus committed:
Reorder the files to create jar
